cpp/src/dataframe/types.hpp

```diff
@@ -38,74 +39,161 @@
+  value_type value;                                    ///< The wrapped value
+
+  CUDA_HOST_DEVICE_CALLABLE
+  explicit wrapper(T v): value{v} {}
+
+  CUDA_HOST_DEVICE_CALLABLE
+  wrapper(wrapper const& w): value(w.value) {}
+
+  CUDA_HOST_DEVICE_CALLABLE
+  wrapper(): value{value_type(0)}{}
+
+  CUDA_HOST_DEVICE_CALLABLE
+  explicit operator value_type() const { return this->value; }
+
+  CUDA_HOST_DEVICE_CALLABLE
+  wrapper& operator=(wrapper const &w)
+  {
+    this->value = w.value;
+    return *this;
+  }
+
+  CUDA_HOST_DEVICE_CALLABLE
+  wrapper& operator+=(wrapper const &w)
   {
-    static constexpr gdf_dtype element_type_id{GDF_CATEGORY};
-    using value_type = gdf_category;
-    value_type value;
-  };
+    this->value += w.value;
+    return *this;
+  }
 
-  struct timestamp
+  CUDA_HOST_DEVICE_CALLABLE
+  wrapper& operator-=(wrapper const &w)
   {
-    static constexpr gdf_dtype element_type_id{GDF_TIMESTAMP};
-    using value_type = gdf_timestamp;
-    value_type value;
-  };
+    this->value -= w.value;
+    return *this;
+  }
 
-  struct date32
+  CUDA_HOST_DEVICE_CALLABLE
+  wrapper& operator*=(wrapper const &w)
   {
-    static constexpr gdf_dtype element_type_id{GDF_DATE32};
-    using value_type = gdf_date32;
-    value_type value;
-  };
+    this->value *= w.value;
+    return *this;
+  }
 
-  struct date64
+  CUDA_HOST_DEVICE_CALLABLE
+  wrapper& operator/=(wrapper const &w)
   {
-    static constexpr gdf_dtype element_type_id{GDF_DATE64};
-    using value_type = gdf_date64;
-    value_type value;
-  };
+    this->value /= w.value;
+    return *this;
+  }
 
-  namespace detail
+  CUDA_HOST_DEVICE_CALLABLE
+  bool operator==(wrapper const &w) const { return this->value == w.value; }
+
+  CUDA_HOST_DEVICE_CALLABLE
+  bool operator!=(wrapper const &w) const { return this->value != w.value; }
+
+  CUDA_HOST_DEVICE_CALLABLE
+  bool operator<=(wrapper const &w) const { return this->value <= w.value; }
+
+  CUDA_HOST_DEVICE_CALLABLE
+  bool operator>=(wrapper const &w) const { return this->value >= w.value; }
+
+  CUDA_HOST_DEVICE_CALLABLE
+  bool operator<(wrapper const &w) const { return this->value < w.value; }
+
+  CUDA_HOST_DEVICE_CALLABLE
+  wrapper operator+(wrapper const &w) const { return wrapper(this->value + w.value); }
+
+  CUDA_HOST_DEVICE_CALLABLE
+  wrapper operator-(wrapper const &w) const { return wrapper(this->value - w.value); }
+
+  CUDA_HOST_DEVICE_CALLABLE
+  wrapper operator*(wrapper const &w) const { return wrapper(this->value * w.value); }
+
+  CUDA_HOST_DEVICE_CALLABLE
+  wrapper operator/(wrapper const &w) const { return wrapper(this->value / w.value); }
+
+  friend std::ostream &operator<<(std::ostream &os, wrapper<T, type_id> const &w)
   {
-    /* --------------------------------------------------------------------------*/
-    /** 
+    return os << w.value;
+  }
+};
+} // namespace detail
+
+struct category : detail::wrapper<gdf_category, GDF_CATEGORY>
+{
+};
```

Comment:
Why should category inherit from `detail::wrapper`  rather than being a `using` alias for it?

Same comment for other empty structs here.

Reply:
It was done to allow for future extensibility where per-wrapper member variables/functions may need to be added.

Reply:
Hmmm... ok.

Reply:
I think you're right. That main reason they are still structs is that that is what they started out as and I didn't think about fully updating them to be an alias for a wrapper rather than an empty struct.

Reply:
Done.